tests/utils/test_proto_json_utils.py

```diff
@@ -569,3 +571,53 @@ def test_dataframe_from_json():
 )
 def test_datetime_encoder(dt, expected):
     assert json.dumps(dt, cls=_CustomJsonEncoder) == expected
+
+
+@pytest.mark.parametrize(
+    ("dataframe", "schema", "expected"),
+    [
+        (
+            pd.DataFrame(columns=["foo"], data=[1, 2, 3]),
+            Schema([TensorSpec(np.dtype("float64"), [-1], "foo")]),
+            np.dtype("float64"),
+        ),
+        (
+            pd.DataFrame(columns=["foo"], data=[[[1, 2, 3]], [[4, 5, 6]]]),
+            Schema([TensorSpec(np.dtype("float64"), [-1, 1], "foo")]),
+            np.dtype("object"),
+        ),
+        (
+            pd.DataFrame(index=[1, 2, 3], columns=["foo"], data=[1, 2, 3]),
+            Schema([TensorSpec(np.dtype("float64"), [-1], "foo")]),
+            np.dtype("float64"),
+        ),
+        (
+            pd.DataFrame(columns=["foo"], data=[1, 2, 3]),
+            Schema([ColSpec("double", "foo")]),
+            np.dtype("float64"),
+        ),
+    ],
+)
+def test_cast_df_types_according_to_schema_success(dataframe, schema, expected):
+    casted_pdf = cast_df_types_according_to_schema(dataframe, schema)
+    assert casted_pdf["foo"].dtype == expected
+
+
+@pytest.mark.parametrize(
+    ("dataframe", "schema", "error_message"),
+    [
+        (
+            pd.DataFrame(columns=["foo"], data=[1, 2, 3]),
+            Schema([ColSpec("binary", "foo")]),
+            r".*TypeError\('encoding without a string argument'\).*",
```

Comment:
```suggestion
            r"TypeError\('encoding without a string argument'\)",
```

I think we can remove `.*`